### ISSUE TITLE
Get stuck in DynamicBucketing Sampler

### ISSUE THREAD

**Issue by csukuangfj** [lhotse-speech/lhotse#736]
I am refactoring icefall to use lazy cutset with dynamic bucketing sampler everywhere in https://github.com/k2-fsa/icefall/pull/397

For the following command run in the librispeech recipe directory using the above PR:

```bash
  ./pruned_transducer_stateless3/train.py \
    --world-size 1 \
    --num-epochs 30 \
    --start-epoch 0 \
    --exp-dir pruned_transducer_stateless3/exp \
    --full-libri 0 \
    --max-duration 100 \
    --giga-prob 0.2
```

the training process seems to get stuck inside the dynamic bucketing sampler.

The log output is
![Screen Shot 2022-06-05 at 23 26 18](https://user-images.githubusercontent.com/5284924/172057948-ee8b80f2-5271-4bdf-b3ff-d5bd13ebda78.png)

I am using `py-spy` to find where it gets stuck:
```bash
watch -n 0.5 py-spy dump --pid 308949 --native
```

The output is

https://user-images.githubusercontent.com/5284924/172058000-fa268eca-5139-4edd-982f-7c0de189bb55.mov

10 minutes have passed but nothing changes.

PS: I am using the latest master of lhotse.

**Comment by pzelasko** [comment]
It seems to be doing a lot of work deserializing supervisions during the estimation of bucket duration bins. Do the supervisions contain alignments (the `AlignmentUnit` objects)? I think that API turned out to be quite inefficient... maybe we need to remove the alignments from Libri recipe (unless you find them useful).

Otherwise it shouldn't take this long with default settings, but you can try to set `num_cuts_for_bins_estimate` and `shuffle_buffer_size` to sth low like 100 just to confirm the script can exit this stage.

**Comment by pzelasko** [comment]
... also did you see a long start-up time for the training script previously? It's suspicious this problem only appeared with dynamic bucketing.

**Comment by csukuangfj** [comment]
Maybe it is my mistake. I am passing a lazy cutset (250 hours) without pre-computed features. 

When I pass a lazy cutset with 10k hours with pre-computed features,  it starts up very quickly.

---

 > Do the supervisions contain alignments (the AlignmentUnit objects)?

No, it does not contain alignment information. It is the gigaspeech S (without pre-computed features) and XL (with pre-computed features) sub datasets.

**Comment by csukuangfj** [comment]
> ... also did you see a long start-up time for the training script previously? It's suspicious this problem only appeared with dynamic bucketing.

No, I have not seen it before.  I have been passing a lazy cutset with precomputed features to the dynamic bucketing samplers. This is the first time I give it a lazy cutset without pre-computed features.

**Comment by csukuangfj** [comment]
I am closing it since it works with lazy cutset with pre-computed features.

**Comment by pzelasko** [comment]
> No, I have not seen it before. I have been passing a lazy cutset with precomputed features to the dynamic bucketing samplers. This is the first time I give it a lazy cutset without pre-computed features.

That's surprising, on-the-fly feature extraction would be happening much later, after the initialization of sampler. Is it possible you had extremely slow I/O/unavailable disk during reading the other manifest...? I can't think of a different explanation.

**Comment by csukuangfj** [comment]
No, the disk is not slow.

I just tried again. After applying `cutset.trim_to_supervisions(keep_overlapping=False, min_duratio=None)`, the cutset without pre-computed features won't get stuck in the dynamic bucketing sampler. This time it fails in `collate_features()`, which is expected since it does not have pre-computed features.

The following shows the statistics of the cutset.

## before using trim_to_supervisions
```
Cuts count: 79
Total duration (hours): 44.4
Speech duration (hours): 10.0 (22.5%)
***
Duration statistics (seconds):
mean    2022.5
std     2588.1
min     202.4
25%     633.7
50%     1330.2
75%     2712.4
99%     14499.3
99.5%   16317.8
99.9%   17772.5
max     18136.2
```

## after applying trim_to_supervisions
```
Cuts count: 9389
Total duration (hours): 10.0
Speech duration (hours): 10.0 (100.0%)
***
Duration statistics (seconds):
mean    3.8
std     2.7
min     0.5
25%     1.9
50%     3.1
75%     5.0
99%     13.9
99.5%   15.6
99.9%   18.5
max     19.9
```

**Comment by pzelasko** [comment]
Ahh, now it makes sense. These cuts are very large and have a lot of supervisions, if you try to load 10k+20k=30k of them, it will take time. You can call `.trim_to_supervisions()` just before passing into the sampler though, I think I made it work with lazy manifests some time ago (if it doesn't, it's a quick fix from my side).

**Comment by csukuangfj** [comment]
> Ahh, now it makes sense. These cuts are very large and have a lot of supervisions, if you try to load 10k+20k=30k of them, it will take time. You can call `.trim_to_supervisions()` just before passing into the sampler though, I think I made it work with lazy manifests some time ago (if it doesn't, it's a quick fix from my side).

Yes. It works for lazy cutset.

**Comment by danpovey** [comment]
I notice that with DynamicBucketingSampler, the validation set on Librispeech reports significantly more frames:
with BucketingSampler:
``` Epoch 14, validation: loss=0.1643, simple_loss=0.2682, pruned_loss=0.03015, over 706207.00 frames.
```
With DynamicBucketingSampler (after some code changes):
```
 Epoch 14, validation: loss=0.1697, simple_loss=0.2705, pruned_loss=0.03449, over 930446.00 frames.
```
... I'm a bit confused, because the number of frames shouldn't depend on how the validation set is arranged into
buckets.

**Comment by danpovey** [comment]
That was with an older Lhotse.  With the latest version of Lhotse, I have verified that the number of frames does not depend on the number of workers, it is 944034.00.   But it's still confusing to me, because from `lhotse   cut describe data/fbank/cuts_dev-clean.json.gz `, and same for dev-other, it seems the total duration of valid set is 5.4+5.1 hours; multiplying by 3600 seconds per hour and 25 frames per second, that should give 945000 frames.  So the length seems correct; I suppose what may have been happening before is the BucketingSampler may have been discarding some utterances.

**Comment by pzelasko** [comment]
That’s possible, I recall merging 2 PRs fairly recently that were fixing some data loss in BucketingSampler.
